### PR TITLE
Hash sensitive data in resource logs

### PR DESCRIPTION
**Pull request by omrirh**
- Hash resource sensitive data in CRUD methods log messages

### CODE REVIEW

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/resource.py [hunk 1]
```diff
@@ -722,8 +723,9 @@ def delete(self, wait=False, timeout=TIMEOUT_4MINUTES, body=None):
         self.logger.info(f"Delete {self.kind} {self.name}")
         if self.exists:
             data = self.instance.to_dict()
-            self.logger.info(f"Deleting {data}")
-            self.logger.debug(f"\n{yaml.dump(data)}")
+            if self.kind != "Secret":
```
[the remark this answers is not in the document]
Comment (omrirh) [reply ocp_resources/resource.py]: done

File: ocp_resources/resource.py [hunk 2]
```diff
@@ -1017,6 +1030,26 @@ def to_yaml(self):
...
+        for key in nested_dict.keys():
+            if key in self.keys_to_hash:
+                if isinstance(nested_dict[key], dict):
+                    nested_dict[key] = {_key: "***" for _key in nested_dict[key].keys()}
+                else:
+                    nested_dict[key] = "***"
+            else:
+                nested_dict[key] = self.hash_resource_dict_rec(nested_dict[key])
+
+        return nested_dict
```
Comment (rnetser) [inline ocp_resources/resource.py:1051]: should be saved in `self.hashed_resource_dict` and not called every time this is needed

File: ocp_resources/resource.py [hunk 3]
```diff
@@ -1017,6 +1030,26 @@ def to_yaml(self):
         self.logger.info(f"\n{resource_yaml}")
         return resource_yaml
 
+    def hash_resource_dict(self, resource_dict: dict):
+        if hasattr(self, "keys_to_hash"):
+            return self.hash_resource_dict_rec(nested_dict=copy.deepcopy(resource_dict))
+        return resource_dict
+
+    def hash_resource_dict_rec(self, nested_dict: dict):
```
Comment (rnetser) [inline ocp_resources/resource.py:1038]: should be in one function

File: ocp_resources/resource.py [hunk 4]
```diff
@@ -1017,6 +1030,26 @@ def to_yaml(self):
...
+
+        for key in nested_dict.keys():
+            if key in self.keys_to_hash:
+                if isinstance(nested_dict[key], dict):
+                    nested_dict[key] = {_key: "***" for _key in nested_dict[key].keys()}
+                else:
+                    nested_dict[key] = "***"
+            else:
+                nested_dict[key] = self.hash_resource_dict_rec(nested_dict[key])
+
```
Comment (rnetser) [inline ocp_resources/resource.py:1050]: no need for this; keys_to_hash should be a list of absolute paths, e.g ["spec.data", "metadata.name"]
Then use python-benedict to get the value from the dict and hash it; no need for recursive function.

File: ocp_resources/secret.py [hunk 5]
```diff
@@ -73,3 +73,10 @@ def certificate_not_before(self):
         return self.instance.metadata.annotations[
             "auth.openshift.io/certificate-not-before"
         ]
+
+    @property
+    def keys_to_hash(self):
+        """
+        Resource attributes to hash in the logs
```
Comment (rnetser) [inline ocp_resources/secret.py:80]:
```suggestion
        Resource attributes list to hash in the logs.
        
         The list should hold absolute key paths in resource dict.
         
         Example:
             ["spec.data"]
```

File: ocp_resources/secret.py [hunk 6]
```diff
@@ -73,3 +73,10 @@ def certificate_not_before(self):
...
+        """
+        return ["data"]
```
Comment (rnetser) [inline ocp_resources/secret.py:82]:
```suggestion
        return ["spec.data"]
```

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/resource.py (hunk 4 above)
Comment (omrirh) [reply ocp_resources/resource.py:1050]: benedict does not support dictionary keys that contain `"."` character:
```
ValueError: Key should not contain keypath separator '.', found: '.dockerconfigjson'.
```

File: ocp_resources/resource.py (hunk 3 above)
Comment (omrirh) [reply ocp_resources/resource.py]: done

File: ocp_resources/resource.py (hunk 4 above)
Comment (omrirh) [reply ocp_resources/resource.py]: done

File: ocp_resources/secret.py (hunk 5 above)
Comment (omrirh) [reply ocp_resources/secret.py]: done

File: ocp_resources/resource.py (hunk 2 above)
Comment (omrirh) [reply ocp_resources/resource.py]: currently for each CRUD method logs we pass a different data dict: Resource `delete` logs `self.instance.to_dict()`, `update` logs a given `resource_dict` and `create logs `self.res`.
Should we use a single attribute of resource as `self.hashed_resource_dict` (e.g. hash `self.instance.to_dict()) ?

**Review by myakove — DISMISSED** [review]

**Review by myakove — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Pull request merged**